optimade/models/structures.py

```diff
@@ -842,12 +877,10 @@ def ratios_must_sum_to_one(cls, v):
 
     @validator("nperiodic_dimensions")
     def check_periodic_dimensions(cls, v, values):
-        if values.get("dimension_types", []) and v is None:
-            raise ValueError(
-                "nperiodic_dimensions is REQUIRED, since dimension_types was provided."
-            )
```

Comment:
Even though some values can be `null`, I would deem it understood that if some specific fields are provided in a resource, then that will automatically require related fields to also be present, i.e., inherently raising them from `SHOULD` to `MUST`?

Reply:
This could be a bit of a sticking point, yeah. I guess the distinction is between `null` in the *response* and `null` in the *implementation*, but the specification doesn't really make that distinction? If e.g. you hadn't precomputed `nperiodic_dimensions` because you could say, well, all my crystals are 3D so this is redundant, would that not be within the bounds of a "SHOULD" interpretation? 

Going with the current approach makes testing the validation significantly simpler, but I'm aware that isn't the best argument... I don't think I was in favour of the near-blanket "SHOULD"-ening of the entire specification, but I think we have to interpret it quite loosely to be consistent with it... We could discuss this tomorrow if we have time

Reply:
Response vs. implementation. That's of course the two central use cases for these models, indeed.

Hmm, well there are still MUST level requirements of inter-field dependencies. This is what I am now seeing being removed. Which to me seems like we're moving away from the specification.

While this is much more easily obvious for the response use case, it should be equally as valid for the implementation, since it by virtue of the specification shouldn't be allowed to serve a resource that doesn't comply with the requirements of the specification.

Hence, there should be two checks for some fields, as there was originally for the highlighted validation, namely both a check of whether the current value is `None` and/or if an inter-field requirement is respected.

I understand the SHOULD level of all fields and their allowance to be `None`/`null` to exist only to the limit of it not violating these inter-field dependencies/requirements or similar. I.e., if you NEED a particular field to be `None`, then it's up to you as the creator of this resource to ensure that all fields relying on this field on a MUST level are also indeed `None`. Since this is the only way all requirements can be fulfilled.
If you wish to save something "wrong" in your private database that's fine, but then it doesn't comply with the OPTIMADE specification and hence it should fail the validation.